NEW: Stock movement pdf & odt document generation

Tim-Otte opened the pull request:
In this PR I fixed the pdf document generation for stock movements and added the odt generation functionality with new custom object tags (object_description, object_nb_products_unique, object_nb_products_total, object_last_movement) and line tags (line_id, line_product_ref, line_qty, line_date, line_label, line_inventory_code, line_price, line_author). The document models for stock movements can be modified and changed in a new tab in the stock module admin page (The tab title is currently only available in english and german).

This is the first atomic part of PR #13677

Commit:
Fixed default pdf template
Added a few missing translations
Added default odt templates for stock and stock movement
Added a new tab to the admin page

stickler-ci committed:
Fixing style errors.

Tim-Otte commented:
The unit test for warehouses (entrepot) and stock movements (mouvementStock) have bugs which are the reason for the failure of the Travis CI Build. This bugs will be fixed with PR #13703 and #13712

eldy commented:
Can you push PR that are "atomic" (one smallest change feature enhancement per PR).
For example, renaming dir doctemplates/stocks/mouvement into doctemplates/stocks/movement should be only 1 PR. So missing thinks on a PR (for example migration on a feature) will not block other changes that are on other PR.

Tim-Otte commented:
1. Renaming the folder for stock movement document templates is atomic, because this directory did not exist before and since if I already fixed the module, I can choose the path myself. Until now, generating documents for stock movements only generated errors. [Here](https://github.com/Dolibarr/dolibarr/blob/5756e3be09e2d41fb367559fe597009218941993/htdocs/core/modules/modStock.class.php#L366) you can see, that no folder was created when initializing the module.
2. There was only one change that could not be atomic: I moved the file pdf_stdmovement.modules.php into the subfolder movement. But since the generation of the documents did not work from the beginning, the file can also be moved to create a clearer structure. This also creates a clear separation of document generators for warehouses and stock movements.
3. Currently there is only one PR with the keyword "[entrepot](https://github.com/Dolibarr/dolibarr/pulls?q=is%3Apr+is%3Aopen+entrepot)" (which is mine) and eight with the keyword "[warehouse](https://github.com/Dolibarr/dolibarr/pulls?q=is%3Apr+is%3Aopen+warehouse)" (one is mine, the other seven do not conflict with my changes, they do not even change the same files). This means that if you handle my PR in time, there will be no conflicts and my changes can be implemented without problems.
4. There is not difference when creating one PR which edits and moves a file or two PRs where one edits the file and a second one moves it. Git automatically tracks [file movements](https://stackoverflow.com/questions/2314652/is-it-possible-to-move-rename-files-in-git-and-maintain-their-history).

@eldy Please merge the PR or give me one good reason to modify my PR.

Tim-Otte committed:
Merge pull request #1 from Dolibarr/develop

Merge changes

eldy commented on htdocs/product/stock/movement_card.php:
```diff
@@ -423,7 +431,7 @@
     $permissiontoread = $user->rights->stock->lire;
     $permissiontodelete = $user->rights->stock->supprimer;
-    $uploaddir = $conf->stock->dir_output."/movement/";
+    $uploaddir = $conf->stock->movement->dir_output;
```
It seems this var does not exist. We try to merge all 
dir_output and dir_temp into 1 dire depending on the $conf->module and then adding a suffix like 'movement'. So code will be ready for hashing directory trees in a future.
Old usage was better.

Tim-Otte replied:
The variable does exist: https://github.com/Dolibarr/dolibarr/blob/39bb69eb87e8d4b5d0ef3cdf59fa06084b668340/htdocs/core/class/conf.class.php#L437
It is part of this PR

Tim-Otte replied:
I changed this to be more compatible to other modules and their configurations. E.g. supplier orders and supplier invoices are configured the same way: https://github.com/Dolibarr/dolibarr/blob/39bb69eb87e8d4b5d0ef3cdf59fa06084b668340/htdocs/core/class/conf.class.php#L386

eldy commented:
Currently this PR contains changes for different goals (refactoring by renaming) and new features. 
It appears this PR can be split into several other smaller PR. We must never mix such PR into 1 PR. It is difficult for people that does not know your code to understand the goal of the PR so to analyze and provides feeback, above all if PR has several goals. Such PR are often never analyzed by other developers that don't want to lose time, and a PR not analyzed, is a PR that we must keep open  a very long time to increase the chance to have an external developer analyzing it. 
It seems there is a misunderstanding of what is an atomic PR. It has no relation with the fact that there is other PR or no done by other on same topic. If they are, it is not a problem. When a PR is atomic, solving conflict can easily be done. Atomic PR is above all to make contributions cleaner for everybody, easier to test before merged, easier to report feedbacks by other people. We also don't mind if the code awas already used or is new for community rules. There is not "one rule for old code" and "another rule for code never used" (this may be true for unit test, but not for atomicity of PR). 
"One PR, One concern" is above all a community working policy. https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1

So can you split your PR into atomic PR ?

eldy commented on htdocs/product/stock/movement_card.php:
```diff
@@ -49,7 +49,7 @@
 $result = restrictedArea($user, 'stock');
 
-$id = GETPOST('id', 'int');
+$entrepot_id = GETPOST('id', 'int');
```
It appears that the movement_card.php is a page view dedicated for 1 warehouse. So id is a werehouse and like for other pages the main object of the page is represented with $id for the id and $object for the main page. It allows to use shared common include files. What is the goal of renaming here into another var ?

Tim-Otte replied:
I changed the name of the variable because the central point of this page is the stock movement and not the warehouse itself. Also, several variables were previously used for different objects and renaming some variables makes it easier to understand the source code.

Tim-Otte commented:
Okay, now I understand what you mean. But I don't know how to split this PR into smaller (atomic) portions of code. All changes in this PR have the goal to fix the document generation for stock movements and to add the functionality to generate odt files for the stock movements. I could split that into "fixing pdf generation" and "adding odt functionality", but both need the fixes and additions to work properly.

Sure, I changed some whitespaces and changed comments and variable names, but only in files I had to change for other reasons. And like I said in my comment above, some variable names had to be changed for better code quality and readability.

I will split the PR if you help me to divide the PR into reasonable atomic PRs that are independent from the other PRs. But currently I don't know how to split it into more atomic PRs.
